lib/path-design.md

```diff
@@ -0,0 +1,384 @@
+# Path library design
+
+This document documents why the `lib.path` library is designed the way it is.
+
+The purpose of this library is to process paths. It does not read files from the filesystem.
+It exists to support the native Nix path value type with extra functionality.
+
+Since the path value type implicitly imports paths from the "eval-time system" into the store,
+this library explicitly doesn't support build-time or run-time paths, including paths to derivations.
+
+Overall, this library works with two basic forms of paths:
+- Absolute paths are represented with the Nix path value type. Nix automatically normalises these paths.
+- Relative paths are represented with the string value type. This library normalises these paths as safely as possible.
+
+Notably absolute paths in a string value type are not supported, the use of the string value type for relative paths is only because the path value type doesn't support relative paths.
+
+This library is designed to be as safe and intuitive as possible, throwing errors when operations are attempted that would produce surprising results, and giving the expected result otherwise.
+
+This library is designed to work well as a dependency for the `lib.filesystem` and `lib.sources` library components. Contrary to these library components, `lib.path` is designed to not read any paths from the filesystem.
+
+This library makes only these assumptions about paths and no others:
+- `dirOf path` returns the path to the parent directory of `path`, unless `path` is the filesystem root, in which case `path` is returned
+  - There can be multiple filesystem roots: `p == dirOf p` and `q == dirOf p` does not imply `p == q`
+    - While there's only a single filesystem root in stable Nix, the [lazy trees PR](https://github.com/NixOS/nix/pull/6530) introduces [additional filesystem roots](https://github.com/NixOS/nix/pull/6530#discussion_r1041442173)
+- `path + ("/" + string)` returns the path to the `string` subdirectory in `path`
+  - If `string` contains no `/` characters, then `dirOf (path + ("/" + string)) == path`
+  - If `string` contains no `/` characters, then `baseNameOf (path + ("/" + string)) == string`
+- `path1 == path2` returns true only if `path1` points to the same filesystem path as `path2`
+
+Notably we do not make the assumption that we can turn paths into strings using `toString path`.
```

Comment:
I updated the PR with such a `commonAncestry` function now (just picked @roberth's recent suggestion for now, I don't have a strong opinion on the name yet).

I also updated the PR to lay down the assumptions about the path type that the library is based on. Notably also including one small assumption which makes it harmonize with the eventual [lazy trees](https://github.com/NixOS/nix/pull/6530) PR.